dotnet/Utils/RichPath.cs

```diff
@@ -0,0 +1,439 @@
+            inStream.Seek(0, SeekOrigin.Begin);
+            return signature[0] == 0x1f && signature[1] == 0x8b;
+        }
+    }
+
+    public class LocalPath : RichPath
+    {
+        public override string Path { get; }
+
+        public LocalPath(string path)
+        {
+            Path = path;
+        }
+
+        public override string ToString() => Path;
+
+        public override bool IsDirectory() => Directory.Exists(Path);
+
+        public override bool IsFile() => File.Exists(Path);
+
+        public override void MakeAsDirectory() => Directory.CreateDirectory(Path);
+
+        public override Stream ReadAsBinaryPotentiallyUncompress()
+        {
+            var inStream = File.OpenRead(Path);
+
+            if (StreamIsGZipped(inStream))
+            {
+                return new GZipStream(inStream, CompressionMode.Decompress);
+            }
+
+            return inStream;
+        }
+
+        public override void Store(object data)
+        {
+            using (var outStream = new FileStream(Path, FileMode.Create))
+            {
+                string normalizedPath;
+                Stream possiblyCompressedStream;
+                if (Path.EndsWith(".gz"))
+                {
+                    normalizedPath = Path.Substring(0, Path.Length - 3); // Remove the ".gz"
+                    possiblyCompressedStream = new GZipStream(outStream, CompressionMode.Compress);
+                }
+                else
+                {
+                    normalizedPath = Path;
+                    possiblyCompressedStream = outStream;
+                }
+
+                if (normalizedPath.EndsWith(".json"))
+                {
+                    StoreAsJSON(possiblyCompressedStream, data, jsonl: false);
+                }
+                else if (normalizedPath.EndsWith(".jsonl"))
+                {
+                    StoreAsJSON(possiblyCompressedStream, data, jsonl: true);
+                }
+                else
+                {
+                    throw new ArgumentException($"File suffix must be .json[.gz] or .jsonl[.gz], but got {Path}");
+                }
+            }
+        }
+
+        public override IEnumerable<RichPath> FilesInDirectory(string searchPattern = "*")
+        {
+            foreach (var newPath in Directory.EnumerateFiles(Path, searchPattern))
+            {
+                yield return new LocalPath(newPath);
+            }
+        }
+
+        public override RichPath Join(string newComponent) 
+            => new LocalPath(System.IO.Path.Combine(Path, newComponent));
+    }
+
+    public class AzurePath : RichPath
+    {
+        public override string Path { get; }
+        private readonly string accountName;
+        private readonly string containerName;
+        private readonly string cacheLocation;
+        private readonly CloudBlobContainer blobContainerClient;
+
+        public AzurePath(string azureInfoPath, string accountPath)
+        {
```

Comment:
I am wondering if a better pattern would be the following:

A special `AzurePathConfig` class encapsulates all azure-related path. This can be created with a the `accountPath`. Then this is passed around to `AzurePath`s.

Although this is minor, this would avoid multiple parsing of the same account path, as well as multiple instantiations of `CloudBlobContainer`.

Reply:
Not done. This is only ever created when directly passing the azureInfoPath (i.e., for example for different specified paths for train/test/valid), but the large number of intermediate paths generated by `FilesInDirectory` uses the copy-constructor that share the `CloudBlobContainer` bits.